### ISSUE TITLE
[INITIATIVE] Fabric8 Kubernetes Client 7.0.0

### ISSUE THREAD

**Issue by manusa** [fabric8io/kubernetes-client#5778]
## Description

Fabric8 Kubernetes Client v7.0.0 major version.

The latest changes in the MockWebServer require a fair deal of breaking changes that must be delivered in a new `x` version to avoid confusion for users.

We're using this opportunity to include as many new features that require breaking changes as possible.

However, going from v6.x to v7.x should not be a very complicated tasks.
Most of the breaking changes just involve replacing imports and only affect a very narrow set of modules.

There are also a minimal set of behavioral changes in the client.

Everything is documented in the [Migration from 6.x to 7.x](https://github.com/fabric8io/kubernetes-client/blob/main/doc/MIGRATION-v7.md) guide.

### Tasks
- #6130 
- #5719 
- #6516 
- #6602 
- #6616 
- #6081 
  - #5965 
  - #6035 
  - #6155 
- #4445
- #6140
- #2632
- #6470 
- #6361 
- #6034

**Comment by jorsol** [comment]
Just one small comment: Bump Java requirement to at least 11, we are in 2024, Java 8 was released 10 years ago, and there are 3 LTS released since Java 8 (11, 17, and 21), so there is no point in supporting Java 8 anymore (unless there is some kind of paid support 😉).

A major version release like 7.0.0 🏷️  is a perfect time to do this kind of change.

I would be more aggressive and bump to Java 17 🚀 , but that is just my personal preference.

**Comment by jorsol** [comment]
> That said, I wouldn't fully agree with you regarding your statement and this project (and for other libraries in general). It's precisely this "there is some kind of paid support" involved in some of the downstream usage of this project _and other libraries_ that make/force these kind of projects the last to update their Java baseline. Apparently Spring 5 (Boot 2) paid support can be at least extended to 2025 🤦.

Well, I didn't explain myself correctly, and haven't mentioned downstream usage, I was referring to paid support to you (or the fabric8 team) not to paid support to "other" projects 😉, and even then, the commercial support should be just for the 6.x branch of KC... the 7.x branch should set the baseline to at least Java 11 without a doubt.

I fully agree that this project (and for other libraries in general), shouldn't be tied to other projects (especially if those other projects are in commercial support), Oracle will have extended support for Java 8 until 31 Dec 2030 (some vendors even longer than that 🤯), it doesn't make any sense that libraries get stuck just because of that.

> The motivation in this case is that is not sustainable any more to keep up with Java 8 (at least not worth the effort). However, this means that we'll probably get an increased amount of requests for patch releases for version 6 (which will be the last one with Java 8 support). I do hope I'm wrong here and that this is not the case but only time will tell.

As mentioned, version 6.x should enter a kind of paid support maintenance, unless there is a critical security issue, there is no point in maintaining older versions, if someone requests for instance support for a newer version of K8s in the 6.x branch, that should be paid support.

I do know that this kind of project makes it hard to provide commercial support, so, in case that is not possible, just provide a best-effort (community-driven) patch release for version 6.

Hope I have cleared the misunderstanding. 😅

**Comment by manusa** [comment]
>  the ~commercial~ best-effort support should be just for the 6.x branch of KC... the 7.x branch should set the baseline to at least Java 11 without a doubt.

Just in case it's not clear, this is what's going to happen exactly :).

I did understand your points, however, I think/feel I didn't explain myself correctly.

For me it's just a matter of tradeoffs. And this project is quite complex in this regard.
On the one hand you have Kubernetes that keeps evolving at a "fast" pace with releases every 6 months. On the other you have dependent projects, libraries, and frameworks that get stuck on a given version of this library.
For me the goal is to keep things as stable as possible and as cross-compatible as possible so that there's no need to maintain multiple branches (major versions) of the project (which would likely be impossible with the amount of resources we have).
Keeping the lowest possible java baseline while things stay simple enough is one of those tradeoffs.
I suppose folks maintaining other popular libraries and tools such as JUnit face similar conundrums.

**Comment by jorsol** [comment]
> For me it's just a matter of tradeoffs. And this project is quite complex in this regard. On the one hand you have Kubernetes that keeps evolving at a "fast" pace with releases every 6 months. On the other you have dependent projects, libraries, and frameworks that get stuck on a given version of this library. For me the goal is to keep things as stable as possible and as cross-compatible as possible so that there's no need to maintain multiple branches (major versions) of the project (which would likely be impossible with the amount of resources we have). Keeping the lowest possible java baseline while things stay simple enough is one of those tradeoffs. I suppose folks maintaining other popular libraries and tools such as JUnit face similar conundrums.

I agree, there will be always tradeoffs, probably the most important thing is to be as stable and cross-compatible as possible, if one of the tradeoffs is to drop support for an older Java version to keep things stable and maintainable, then is a very welcome tradeoff.

**Comment by martinwunderlich-celonis** [comment]
Hello all, I am wondering what the expected release date for version 7.0.0 currently is? We'd like to upgrade soon to get the new kubernetes client and get rid of the bouncy-castle dependency. Thanks!

**Comment by manusa** [comment]
We've made a lot of progress and we should be able to release in a couple of weeks.
In the meantime, you should be able to try the snapshots and report any bugs that might have come up while implementing the new features and after the refactoring efforts for the current ones.
If you're on the latest 6.13 release, the upgrade effort should be minimal.
Please, also check the [Migration Guide](https://github.com/fabric8io/kubernetes-client/blob/main/doc/MIGRATION-v7.md) and report anything that might be missing.

**Comment by martinwunderlich-celonis** [comment]
@manusa Thanks a lot for the quick feedback. I have already tested with a local snapshot build and it all works fine. We haven't discovered any bugs and are looking forward very much to the release. :)
